File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        return self.dfs(root, root.val)


    def dfs(self, root: TreeNode, path: int) -> int:
        if not root: return 0
        newPath = max(path, root.val)

        return self.dfs(root.left, newPath) + self.dfs(root.right, newPath) + (1 if root.val >= path else 0) 
        

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    record NodePair (TreeNode node, int max) {};
    public int goodNodes(TreeNode root) {

        int count = 0;
        LinkedList<NodePair> queue = new LinkedList<>();
        queue.offer(new NodePair(root, root.val));

        while (queue.size() > 0)
        {
            NodePair pair = queue.poll();
            TreeNode node = pair.node();

            if (node.val >= pair.max())
            {
                count += 1;
            }

            int newMax = Math.max(node.val, pair.max());

            if (node.left != null) queue.offer(new NodePair(node.left, newMax));
            if (node.right != null) queue.offer(new NodePair(node.right, newMax));
        }

        return count;
    

    }
}